scipy/optimize/_direct.py

```diff
@@ -0,0 +1,263 @@
+           79, 157-181 (1993)
+    .. [2] Gablonsky, J., Kelley, C. A Locally-Biased form of the DIRECT
+           Algorithm. Journal of Global Optimization 21, 27-37 (2001).
+
+    Examples
+    ----------
+    The following example is a 2-D problem with four local minima: minimizing
+    the Styblinski-Tang function
+    (https://en.wikipedia.org/wiki/Test_functions_for_optimization).
+
+    >>> from scipy.optimize import direct, Bounds
+    >>> def styblinski_tang(pos):
+    ...     x, y = pos
+    ...     return 0.5 * (x**4 - 16 * x**2 + 5 * x + y**4 - 16 * y**2 + 5 * y)
+    >>> bounds = Bounds([-4., -4.], [4., 4.])
+    >>> result = direct(styblinski_tang, bounds)
+    >>> result.x, result.fun, result.nfev
+    array([-2.90362242, -2.90362242]), -78.33233113735979, 20003
+
+    The correct global minimum was found but with a huge number of function
+    evaluations (20003). Loosening the termination tolerances `vol_tol` and
+    `len_tol` can be used to stop DIRECT earlier.
+
+    >>> from scipy.optimize import direct, Bounds
+    >>> def styblinski_tang(pos):
+    ...     x, y = pos
+    ...     return 0.5 * (x**4 - 16 * x**2 + 5 * x + y**4 - 16 * y**2 + 5 * y)
+    >>> bounds = Bounds([-4., -4.], [4., 4.])
+    >>> result = direct(styblinski_tang, bounds, vol_tol = 1e-7)
+    >>> result.x, result.fun, result.nfev
+    array([-2.90321597, -2.9044353 ]), -78.33231560853986, 1113
+    """
+
+    # convert bounds to new Bounds class if necessary
+    if not isinstance(bounds, Bounds):
+        if isinstance(bounds, list) or isinstance(bounds, tuple):
+            lb, ub = old_bound_to_new(bounds)
+            bounds = Bounds(lb, ub)
+        else:
+            message = ("bounds must be a sequence or "
+                      "instance of Bounds class")
+            raise ValueError(message)
+
+    lb = np.ascontiguousarray(bounds.lb)
+    ub = np.ascontiguousarray(bounds.ub)
+
+    # validate bounds
+    if not np.all(lb < ub):
+        raise ValueError('Bounds are not consistent min < max')
+    if not len(lb) == len(ub):
+        raise ValueError('Bounds do not have the same dimensions')
+
+    # check for infs and nans
+    if (np.any(np.isinf(lb)) or np.any(np.isinf(ub))):
+        raise ValueError("Bounds must not be inf.")
+    if (np.any(np.isnan(lb)) or np.any(np.isnan(ub))):
+        raise ValueError("Bounds must not be NaN.")
+
+    # validate tolerances
+    if (vol_tol < 0 or vol_tol > 1):
+        raise ValueError("vol_tol must be between 0 and 1.")
+    if (len_tol < 0 or len_tol > 1):
+        raise ValueError("len_tol must be between 0 and 1.")
+    if (f_min_rtol < 0 or len_tol > 1):
+        raise ValueError("f_min_rtol must be between 0 and 1.")
+
+    # validate maxfun and maxiter
+    if not isinstance(maxfun, int):
+        raise ValueError("maxfun must be of type int.")
+    if not isinstance(maxiter, int):
+        raise ValueError("maxiter must be of type int.")
+
+    def _func_wrap(x, *args):
```

Comment:
```suggestion
    def _func_wrap(x):
```
What's probably happening is `args=None` from `_direct`. This will make sure you pass into `func` the `args` passed into `direct`.

Reply:
That's right. Somehow `args` aren't getting relayed to the input function. This should fix that.

Reply:
This does not solve the issue unfortunately. Functions without args still work but args passthrough still fails.

Error message:
`TypeError: direct.<locals>._func_wrap() takes 1 positional argument but 2 were given`

I think the bug lies somewhere in the C wrapper.

Reply:
I see. If it's okay, I can look into the issue tomorrow night and day after tomorrow as I plan to work on SciPy in the next two days. Let me know.

Reply:
Yes that would be great if you can find time, thx!

Reply:
Fixed and added a test.